### PR TITLE
Add CI settings to test Matplotlib without Plotly

### PR DESCRIPTION
**Pull request by DanielAvdar**
## Motivation

Add CI settings to check optuna.visualization.matplotlib module is available without plotly #5205
## Description of the changes

add new workflow for matplotlib tests.

### CODE REVIEW

**Comment by nzw0301** [comment]
I think Optuna repo does not prefer to using `utils.py` as described in https://github.com/optuna/optuna/issues/4593. So reviewers might ask you to revert the change.

**Comment by nabenabe0928** [comment]
@c-bata Could you review this PR when you have time as the original issue was created by you?
The PR review is not super urgent.

**Comment by DanielAvdar** [comment]
> I think Optuna repo does not prefer to using `utils.py` as described in #4593. So reviewers might ask you to revert the change.

This change allows running Matplotlib tests without being dependent on Plotly, which was imported from tests/visualization_tests/test_timeline.py. Generally, I'm open to suggested alternatives.

**Comment by c-bata** [comment]
@nabenabe0928 Thank you for asking. I'm tied up at the moment and given that the changes in this PR are mainly about unit tests and CI settings, it would be great if other maintainers could take a look. Could you please reassign this PR to someone else?

**Comment by nabenabe0928** [comment]
@HideakiImamura Could you review this PR instead of @c-bata ?

**Comment by nzw0301** [comment]
@DanielAvdar Thanks. After reading the existing code a little bit, in my understanding, we can remove `tests/visualization_tests/matplotlib_tests/test_timeline.py` by following the test file for visualisation, e.g. https://github.com/optuna/optuna/blob/cbf62c341acbc33f4355b7b4acd8c99834730d28/tests/visualization_tests/test_slice.py#L101-L112 by performing the test in `tests/visualization_tests/test_timeline.py`.

@nabenabe0928 @HideakiImamura Please correct me if my understanding is wrong.

**Comment by DanielAvdar** [comment]
> @DanielAvdar Thanks. After reading the existing code a little bit, in my understanding, we can remove `tests/visualization_tests/matplotlib_tests/test_timeline.py` by following other test files for visualisation, e.g.
> 
> https://github.com/optuna/optuna/blob/cbf62c341acbc33f4355b7b4acd8c99834730d28/tests/visualization_tests/test_slice.py#L101-L112
> 
> by performing the test in `tests/visualization_tests/test_timeline.py`.
> @nabenabe0928 (and @HideakiImamura ?) Please correct me if my understanding is wrong.

I can remove utils.py, matplotlib_tests/test_timeline.py, and move back the _create_study function to tests/visualization_tests/test_timeline.py if that's desired.

**Comment by DanielAvdar** [comment]
@nzw0301 

> Thanks. As you explained, the master branch's `tests/visualization_tests/matplotlib_tests/test_timeline.py` depends plotly via `tests/visualization_tests/test_timeline.py`. To avoid import error without `plotly`, adding a few lines:
> 
> https://github.com/optuna/optuna/blob/cbf62c341acbc33f4355b7b4acd8c99834730d28/tests/visualization_tests/test_edf.py#L24-L25
> 
> to `tests/visualization_tests/test_timeline.py` is required in addition to removal of the introduced changes.


Here are the changes you asked for, if I understand you correctly.

File: .github/workflows/matplotlib-tests.yml
Comment (nzw0301) [inline .github/workflows/matplotlib-tests.yml:3]:
```suggestion
```

File: .github/workflows/matplotlib-tests.yml
Comment (nzw0301) [inline .github/workflows/matplotlib-tests.yml:9]: Matplotlib-dependent tests exist in `tests/visualization_tests/*.py` too.
```suggestion
      - tests/visualization_tests/**
```

File: .github/workflows/matplotlib-tests.yml
Comment (nzw0301) [inline .github/workflows/matplotlib-tests.yml:25]: I think these version updates proposed in https://github.com/optuna/optuna/pull/5253 haven't been accepted by the master branch yet. Using an older version makes this PR review smooth.

File: .github/workflows/matplotlib-tests.yml
Comment (nzw0301) [inline .github/workflows/matplotlib-tests.yml:40]: In my understanding, this cache is defined for [mac-test](https://github.com/optuna/optuna/blob/master/.github/workflows/mac-tests.yml). Could you replace this with the ubuntu version, e.g., https://github.com/optuna/optuna/blob/master/.github/workflows/tests-with-minimum-versions.yml#L43-L51.

File: .github/workflows/matplotlib-tests.yml
Comment (nzw0301) [inline .github/workflows/matplotlib-tests.yml:69]: I'm not sure these two test executions are necessary Can we remove them?

File: .github/workflows/matplotlib-tests.yml
Comment (nzw0301) [inline .github/workflows/matplotlib-tests.yml:74]:
```suggestion
        pytest tests/visualization_tests
```

**Review by nzw0301 — CHANGES_REQUESTED** [review]

**Comment by nzw0301** [comment]
My main concern is that the current PR does not address the issue completely because many matplotlib's tests exist in `tests/visualization_tests/*.py`, not `tests/visualization_tests/matplotlib_tests`. I think what we need is to make`tests/visualization_tests/` run without `plotly`.

**Comment by DanielAvdar** [comment]
> My main concern is that the current PR does not address the issue completely because many matplotlib's tests exist in `tests/visualization_tests/*.py`, not `tests/visualization_tests/matplotlib_tests`. I think what we need is to make`pytest tests/visualization_tests/` run without `plotly`.

@nzw0301, thank you for your review. I have incorporated the desired changes. Please let me know if I missed anything. It might be possible to consolidate all plotly-dependent tests into one location.

File: .github/workflows/matplotlib-tests.yml
Comment (nzw0301) [inline .github/workflows/matplotlib-tests.yml:62]: By the same as one of the previous my review comments, this seems to be done by another CI job and not related to visualization part.

```suggestion
```

File: .github/workflows/matplotlib-tests.yml
Comment (nzw0301) [inline .github/workflows/matplotlib-tests.yml:32]: Could you rename this since this cache name for the other CI job?

File: .github/workflows/matplotlib-tests.yml
Comment (nzw0301) [inline .github/workflows/matplotlib-tests.yml:8]: Same ordering as the optuna repo would be better.

```suggestion
      - .github/workflows/matplotlib-tests.yml
      - optuna/visualization/**.py
      - tests/visualization_tests/**
```

**Review by nzw0301 — CHANGES_REQUESTED** [review]
Thank you for your update. I've left a few more suggestions.

**Comment by nzw0301** [comment]
Addressing https://github.com/optuna/optuna/pull/5263#issuecomment-1958775671 will need discussion with the core-dev team since it is more little bit more work (e.g. https://github.com/optuna/optuna/commit/0ffa6d516fb255fd0cd631ceca578bada3ed4f6f), so let's wait for a response from them 🤞🏻

File: .github/workflows/matplotlib-tests.yml
Comment (nzw0301) [inline .github/workflows/matplotlib-tests.yml:59]: One black line is enough.
```suggestion

```

**Review by nzw0301 — CHANGES_REQUESTED** [review]
One more thing.

**Review by nzw0301 — APPROVED** [review]
Thank you for your updates again! Except for the pull request title, LGTM. 

Could you re-pharase the pull request title? It will be used in the released note, where an informative and concrete title is helpful.

**Comment by DanielAvdar** [comment]
> Thank you for your updates again! Except for the pull request title, LGTM.
> 
> Could you re-pharase the pull request title? It will be used in the released note, where an informative and concrete title is helpful.

Changed, I'm open to new suggestions.

**Review by HideakiImamura — APPROVED** [review]
LGTM.

**Pull request merged**
